File: Solutions/Source2/Searching/FacingTheSun.py
'''
By: Marcos Gil
Solution for this practice problem
https://practice.geeksforgeeks.org/problems/facing-the-sun/0
'''

'''
Name: printResult
Purpose: Print the result
 In: inSun = int for buildings in sun,
     buildingList = list of building heights
 In/Out: N/A
 Out: N/A
'''
def printResult(inSun, buildingList):
    if inSun < 0:
        print("Empty list was entered")
    else:
        print("The buildings are:", buildingList)
        print("There are", inSun, "buildings receiving sunlight.")

'''
Name: getList
Purpose: Get building heights from user as a list
 In: N/A
 In/Out: N/A
 Out: list = list of building heights
'''
def getList():
    userValues = input("Enter space seperated digits as building height values: ").split(" ")
    list = [int(x) for x in userValues]
    return list

'''
Name: amountInSun
Purpose: Calculate the amount of buildings in the sun
 In: buildings = list of building heights
 In/Out: N/A
 Out: numFacingSun = int value of buildings in sun
'''
def amountInSun(buildings):
    if (len(buildings) == 0):
        return -1
    else:
        numFacingSun = 1
        lastHeight = buildings[0]

        for i in range (1, len(buildings)):
            if buildings[i] > lastHeight:
                numFacingSun += 1
                lastHeight = buildings[i]

        return numFacingSun

def main():
    buildings = getList()
    inSun = amountInSun(buildings)
    printResult(inSun, buildings)

main()

File: Solutions/Source2/Searching/FacingTheSun.java
/*
 * By: Marcos Gil
 * Solution for this practice problem
 * https://practice.geeksforgeeks.org/problems/facing-the-sun/0
 */

import java.util.Scanner;
import java.util.Arrays;

public class FacingTheSun{

  /*
  Name: getArray
  Purpose: Get the array to search
    In: N/A
    In/Out: N/A
    Out: int[] array
  */
  public static int[] getArray(){

    int lengthOfArray = -1;
    Scanner userInput = new Scanner(System.in);
    System.out.print("Please enter the length of the array: ");
    lengthOfArray = userInput.nextInt();

    int[] array = new int[lengthOfArray];

    for (int i = 0; i < lengthOfArray; i++){
      System.out.print("Enter value " + (i + 1) + " of the array: ");
      array[i] = userInput.nextInt();
    }

    return array;
  }

  /*
  Name: facingTheSun
  Purpose: Calculate the amount of buildings facing the sun
    In: int[] buildingHeights
    In/Out: N/A
    Out: int buildingsFacingSun
  */
  public static int facingTheSun(int[] buildingHeights){

    if (buildingHeights.length == 0){
      return -1;
    }

    int buildingsFacingSun = 1;
    int lastHeight = buildingHeights[0];

    for (int i = 1; i < buildingHeights.length; i++){

      if (buildingHeights[i] > lastHeight){

        buildingsFacingSun++;
        lastHeight = buildingHeights[i];
      }
    }

    return buildingsFacingSun;
  }

  public static void main(String[] args) {

    int[] userArray = getArray();
    int numOfBuildingsFacingSun = facingTheSun(userArray);

    if (numOfBuildingsFacingSun == -1){
      System.out.println("Empty array was entered");
    } else {
      System.out.println("The array is: " + Arrays.toString(userArray));
      System.out.println("There are " + numOfBuildingsFacingSun + " buildings receiving sunlight.");
    }
  }
}
